contracts/test/unit/oracle/DynamicVoltSystemOracle.t.sol

```diff
@@ -0,0 +1,327 @@
+            systemOracle.periodStartOraclePrice()
+        );
+
+        // grow at new change rate for half of TIMEFRAME
+        vm.warp(periodStartTime + TIMEFRAME);
+        assertEq(systemOracle.getCurrentOraclePrice(), 1.2075e18);
+
+        // still grow for half of TIMEFRAME because a new period has started
+        vm.warp(periodStartTime + (TIMEFRAME * 3) / 2);
+        assertEq(systemOracle.getCurrentOraclePrice(), 1.365e18);
+    }
+
+    function testUpdateBaseRateLowLiquidity() public {
+        lastLiquidVenuePercentage = 0; // 0% => max boost
+        uint256 maxRate = rateModel.MAXIMUM_CHANGE_RATE();
+
+        vm.warp(periodStartTime);
+
+        // set base rate to 3x the current base rate
+        uint256 newBaseRate = baseChangeRate * 3;
+        // check events
+        vm.expectEmit(false, false, false, true, address(systemOracle));
+        emit InterestCompounded(uint64(periodStartTime), 1e18);
+        vm.expectEmit(false, false, false, true, address(systemOracle));
+        emit BaseRateUpdated(
+            uint64(periodStartTime),
+            baseChangeRate,
+            newBaseRate
+        );
+        vm.expectEmit(false, false, false, true, address(systemOracle));
+        emit ActualRateUpdated(
+            uint64(periodStartTime),
+            baseChangeRate,
+            maxRate
+        );
+        // prank & update base rate
+        vm.prank(addresses.governorAddress);
+        systemOracle.updateBaseRate(newBaseRate);
+
+        // check updated rates
+        assertEq(systemOracle.baseChangeRate(), newBaseRate);
+        assertEq(systemOracle.actualChangeRate(), maxRate);
+    }
+
+    function testUpdateBaseRateAcl() public {
+        vm.expectRevert(bytes("CoreRef: Caller is not a governor"));
+        systemOracle.updateBaseRate(0);
+    }
+
+    function testSetRateModel() public {
+        uint256 currentTimestamp = periodStartTime;
+        vm.warp(currentTimestamp);
+        assertEq(systemOracle.rateModel(), address(rateModel));
+
+        // check events
+        vm.expectEmit(false, false, false, true, address(systemOracle));
+        emit RateModelUpdated(currentTimestamp, address(rateModel), address(0));
+
+        // prank & call
+        vm.prank(addresses.governorAddress);
+        systemOracle.setRateModel(address(0));
+        assertEq(systemOracle.rateModel(), address(0));
+
+        // check events
+        vm.expectEmit(false, false, false, true, address(systemOracle));
+        emit RateModelUpdated(currentTimestamp, address(0), address(rateModel));
+
+        // prank & call
+        vm.prank(addresses.governorAddress);
+        systemOracle.setRateModel(address(rateModel));
+        assertEq(systemOracle.rateModel(), address(rateModel));
+    }
+
+    function testSetRateModelAcl() public {
+        vm.expectRevert(bytes("CoreRef: Caller is not a governor"));
+        systemOracle.setRateModel(address(0));
+    }
+
+    function testUpdateActualRateFuzz(
```

Comment:
would be great to see a test like this that iterates over multiple periods and compounds interest a few times

Reply:
I'm adding a test that fuzz random period durations & tests compounding over 10 periods